Balloting tool doku (#1)
HL7Austria/ballot-comments: docs/balloting-tool-documentation.md

The comment should be visible at the top. To see more details click on the discussion. This will look similar to this [example](https://github.com/HL7Austria/ELGA-FHIR-e-Medikation-R5/discussions/54): 

![discussion example](discussions-example.png)

### !Attention! 
 - Highlights usually appear within 1–5 minutes.
 - Selections should be kept as short as possible. Only select what is relevant to your comment. 
 - Only one paragraph or headline can be selected at a time
 - If a slection gives an error try again with a slightly different selction (can happen if bold and regular text try to be selected together)

# Balloting Tool for IG creators
## Discussions Setup
Follow [this guide](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/enabling-or-disabling-github-discussions-for-a-repository) to enable discussions on the repository that the ballot should happen in.

## Giscus Setup

1. Install github app giscus for your repo using this [link](https://github.com/apps/giscus). Press "configure" and select HL7Austria as the Organization.
<img src="giscus-app.png" alt="giscus app" width="700" >

> 3. Select the repository for which giscus should be enabled and request access.

Author: Rewrite numbering (1,3,5,7,8)

<img src="giscus-repository-access.png" alt="giscus repository access" width="400" >

5. Once access is granted go to this [website](https://giscus.app) and enter the repository name. If steps 1 and two were successful a green checkmark will indicate that giscus is enabled an can be used.
   
![giscus app](giscus-check.png)

> 7. Set the discussion Category to Comment and then copy the following lines:

Author: By default `Comment` does not exist. I assume this category was created by Nik.

If discussions are enabled for the first time in a repo, the following categories are created:
![image](https://github.com/user-attachments/assets/c0779023-e84d-4262-868f-0037e2504cd0)

As a recommendation "General" is a good starting point.

        data-repo="HL7Austria/ELGA-FHIR-e-Medikation-R5"
        data-repo-id="R_kgDOLhyFQQ"
        data-category="Comment"

>         data-category-id="DIC_kwDOLhyFQc4Cev5-"

Author: We should probably include a screenshot here and highlight the lines which are relevant for the next step.

![image](https://github.com/user-attachments/assets/4c368509-3a33-491d-a30c-80dbe88cefc1)

> 8. These values need to be updated in custom-template/content/assets/js/comment-box.js during the IG setup.

Author: ```suggestion
8. The values of `data-repo`, `data-repo-id`, `data-category`, and `data-category-id` have to be copied into a file called `giscus-config.js`. This file has to be placed in your IG unter `input/js`.

    const GITHUB_REPO = 'your-data-repo-value-goes-here';
    const GITHUB_REPO_ID = 'your-data-repo-id-value-goes-here';
    const GITHUB_CATEGORY = 'your-data-category-value-goes-here';
    const GITHUB_CATEGORY_ID = 'your-data-category-id-value-goes-here-';
```

## IG Setup
TBD once established

> Ensure that the file UpdateBallotComments.yml exist in ```.github/workflows```

Author: We probably should put the UpdateBallotComments.yml into this repository.

> Ensure that this line is included towards the end of fragment-pageend.html: ```<script type="text/javascript" src="assets/js/comment-box.js"></script>```

Author: ```suggestion
Ensure that within your `ig.ini` the template is being set to `template = hl7.at.fhir.template.ballot#current`

If you use your own template, please take a look at https://github.com/HL7Austria/at-fhir-ig-template-ballot and set it upon your own template.
```

## Final IG Ballot Checklist
- [ ] GitHub Discussions enabled

> - [ ] "Comment" category exists

Author: ```suggestion
- [ ] "General" category exists
```

- [ ] Giscus app installed & configured

> - [ ] `comment-box.js` updated with correct repo and repo ID

Author: ```suggestion
- [ ] `giscus-config.js` updated with correct repo and repo ID
```

- [ ] `UpdateBallotComments.yml` workflow exists

> - [ ] IG page includes `<script src=".../comment-box.js">`

Author: ```suggestion
```

# Balloting Tool Highlighting Mechanism

Upon creation of a new discussion (by pressing comment in the giscus window) a new discussion is created by the giscus bot. The creation of a new discussion triggers the UpdateBallotComments github action that is defined in ```.github/workflows/UpdateBallotComments.yml``` . 

### discussion id

The balloting tool relies on a unique string — referred to as the discussion ID, giscus term or highlight ID — to link each comment to a specific section of the IG. This string is used by:

Giscus to fetch and identify the correct discussion thread.
The highlighting mechanism to locate and mark the correct passage in the rendered IG.
The term has the following structure:

```[page path] | [compressed XPath] | [startOffset] | [endOffset]```

Each component is separated by a pipe |. For example:

```/r5-ELGA-e-medikation-section-specific-ballot-tool-approach/index.html|/html[1]/body[1]/div[4]/div[1]*4/p[3]/text()|33|85```

To reduce the overall length (and avoid giscus bugs), the XPath portion is compressed using a custom scheme that replaces repeated elements (e.g., /div[1]/div[1]/div[1] becomes /div[1]*3). This compression is essential because:

> Giscus imposes a strict character limit on discussion titles. If this limit is exceeded, giscus silently fails to locate or render the associated discussion, without showing any error message. The compression format is reversible and the GitHub Action described below automatically decompresses it when building the comments.json used by the frontend.

Author: could the strict character limit be a problem sometimes? Would it be possible to use an md5 hash of the string above as discussion id and store the string within the discussion?

> ### UpdateBallotComments Github Action

Author: Would it be possible to determine the branch the comment was created on? Maybe as part of the URL?
